Refactor Documentation (#645)
graalvm/native-build-tools: docs/src/docs/asciidoc/end-to-end-gradle-guide.adoc

        debug = true
        verbose = true
    }

    binaries.all {
        quickBuild = false
    }
}
----

[source,kotlin,subs="verbatim,attributes", role="multi-language-sample"]
----
graalvmNative {
    binaries.main {
        imageName.set('application')
        mainClass.set('org.test.Main')
        fallback.set(true)
        sharedLibrary.set(false)
        richOutput.set(false)
        requiredVersion.set('22.3')

        systemProperties.putAll(mapOf("name1" to "value1", "name2" to "value2"))
        configurationFileDirectories.from(file('src/my-config'))

        buildArgs.add('--link-at-build-time')
        runtimeArgs.add('--help')

        jvmArgs.add('flag')
        useFatJar.set(true)
    }

    binaries.test {
        debug.set(true)
        verbose.set(true)
    }

    binaries.all {
        quickBuild.set(false)
    }
}
----

[[collect-metadata]]
== Collecting metadata

When your test/application starts to be a bit more complex things like **reflection**, **resources**, **serialization**, **proxies** or **jni** may be required.
Since the Native Image has closed world assumption, all of these things must be known in advance during the image build.

> The easiest way how this information can be passed to the Native Image is through metadata config file(s) - depending on the GraalVM version you are using, there could be

Reviewer: ```suggestion
The easiest way to pass this information to the Native Image is through configuration file(s) - depending on the GraalVM version you are using, there could be
```

a single `reachability-metadata.json` file (for newer GraalVM versions) or multiple json files (`reflect-config.json`, `resource-config.json`, `proxy-config.json`, `serialization-config.json`, `jni-config.json`).
To learn more about metadata that Native Image consumes, https://www.graalvm.org/latest/reference-manual/native-image/metadata/[see this].

> For example, if you run the test that uses the reflection, and you don't have entry for that reflection call in the metadata config file, your test will fail.

Reviewer: ```suggestion
For example, if you run tests which make use of reflection, and you don't have entry for reflective calls in the metadata config file, your native tests will fail.
```

> To make your test/application work while using reflection (like in this example) or other metadata, you should either generate metadata configurations or write them manually.

Reviewer: ```suggestion
In order to make your test/application work natively while using reflection (like in this example) or other features which require additional metadata, you should either generate configuration files or write them manually.
```

> To generate metadata automatically, you can run your tests (or the main application) with the Native Image Agent, that will collect all the metadata your test/application require.

Reviewer: ```suggestion
The easiest is to generate metadata automatically.
For this you can run your tests (or the main application) with the Native Image Agent, which will collect all the metadata your test/application require and generate metadata for you.
```

To enable the agent (through Native Gradle Plugin) you should either:

> - add `-Pagent` flag to the command you are executing

Reviewer: ```suggestion
- add `-Pagent` flag to the command line (for example `./gradlew test -Pagent`)
```

- or add the following block to `graalvmNative` block in the `build.gradle`:

[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
----
agent {
    enabled = true
}
----

[source,kotlin,subs="verbatim,attributes", role="multi-language-sample"]
----
agent {
    enabled.set(true)
}
----

[.underline]#To generate the metadata file(s) for your `tests` just run:#

- `./gradlew test` if you added the agent block to the configuration or `./gradlew -Pagent test` if you didn't. This command runs on JVM with native-image-agent and collects the metadata.
- `./gradlew nativeTest` if you added the agent block to the configuration or `./gradlew -Pagent nativeTest` if you didn't. This command runs on JVM with the native-image agent, collects the metadata and uses it for testing on native-image.

[.underline]#To generate the metadata file(s) for your `application` just run:#

- `./gradlew run` if you added the agent block to the configuration or `./gradlew -Pagent run` if you didn't. This command runs on JVM with native-image-agent and collects the metadata.

[WARNING]
====

> Unless you specify the following block in your `build.gradle` file, Gradle will pick up the Agent from the `JAVA_HOME` environment variable.

Reviewer: :question: Are you sure about this? Isn't it the problem that the application may be _compiled_ with one Java version, and that the agent uses the version from `GRAALVM_HOME`? That would be what I expect.

Author: Discussed offline. Please take a look again

> This may cause problems if you set `GRAALVM_HOME` environment variable because your project will be build with one version and the Agent will generate metadata for the other one.

Reviewer: ```suggestion
This may cause problems if you set `GRAALVM_HOME` environment variable because your project will be built with one version and the Agent will generate metadata for the another one.
```

[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
----
executable = providers.environmentVariable("GRAALVM_HOME").map {
    "$it/bin/java"
}.get()
----

This way, generated metadata config file(s) will have format that is specified in the GraalVM version (from `GRAALVM_HOME`) you are using.
====

[[metadata-copy]]

> === Move generated metadata to non-default location

Reviewer: ```suggestion
=== Copy generated metadata to permanent location
```

By default, generated metadata will be placed inside `build/native/agent-output` directory.

> In many cases you may want to move generated metadata to some other location.

Reviewer: ```suggestion
In many cases you may want to copy generated metadata to some other location.
```

To do so, you can configure and run `metadataCopy` task.

==== Configure metadataCopy task

First, you can configure `metadataCopy` task by adding a new block, named `metadataCopy` inside `agent` block that you added in the previous step.
Inside this block, you can specify:

> - `outputDirectories` - location where you want to move the metadata

Reviewer: ```suggestion
- `outputDirectories` - location where you want to copy the generated metadata
```

- __mergeWithExisting = false__ - means that we don't want to merge incoming metadata with the one that already exists on the location specified in `outputDirectories` (this makes sense since we don't have metadata on the given location already)

[[execute-metadata-copy-task]]
==== Execute metadataCopy task

Once the metadata is generated and the `metadataCopy` task is configured, you can run the task with:

[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----
./gradlew metadataCopy
----

Besides that, you can configure `metadataCopy` task through the command line as well:

[source,bash,subs="verbatim,attributes"]
----
./gradlew metadataCopy --task run   # if you used nativeRun (or just run) to collect metadata
./gradlew metadataCopy --task test  # if you used nativeTest (or just test) to collect metadata
./gradlew metadataCopy --dir <pathToSomeDirectory> # to specify the output directory
----

Here is an example of a valid `metadataCopy` usage:

[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----
./gradlew metadataCopy --task test --dir resources/META-INF/native-image/org.example
----

[WARNING]
====

> Note that **if you store generated metadata on location other than the default one**, you will need to pass that location as a Native Image `buildArg` with `-H:ConfigurationFileDirectories` option.

Reviewer: ```suggestion
Note that **if you store generated metadata outside of the standard resources directories (e.g `src/main/resources/META-INF/native-image/...`) **, you will need to pass that location as a Native Image `buildArg` with `-H:ConfigurationFileDirectories` option.
In this case Gradle will _not_ be able to detect changes to metadata properly, which is why we recommend to use standard locations.
```

First thing that you can configure is the agent mode.
There are three possible agent modes:

* `standard` - only generates metadata without any special processing (this is the default mode). No additional options available.
* `conditional` - entries of the generated metadata will be included in the Native Image only if the condition in the entry is satisfied. Consumes following options:
** `userCodeFilterPath` - specifies a filter file used to classify classes as user application classes. Generated conditions will only reference these classes See <<agent-filter-file, the following section>>
** `extraFilterPath` - (optional) extra filter used to further filter the collected metadata. See <<agent-filter-file, the following section>>
* `direct` - in this mode user configures the agent completely manually by adding all options with:
** `options.add("<option>")`

Each of the described modes has its own benefits.
For example:

- `standard` mode is a **great starting point** in your project development
- `conditional` mode is mainly aimed towards **library maintainers** with the goal of reducing overall footprint

> - `direct` mode is for **experienced users** that knows how to configure the agent manually

Reviewer: ```suggestion
- `direct` mode is for **experienced users** who know how to configure the agent manually
```

You can configure each mode (and declare the one that will be used for generating metadata) inside the `agent` block in `build.gradle` file.
Here is an example of the `agent` block with configured conditional and direct modes, where the conditional mode is set as default and will be used to generate the metadata:

[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
----
agent {
    enabled = true
    defaultMode = "conditional"
    modes {
        conditional {

>             userCodeFilterPath = "src/test/resources/metadata.with.starting.user-code-filter/user-code-filter.json"

Reviewer: :nit: Technically speaking, files found in `src/test/resources` are found on classpath at runtime (here, for tests). This doesn't quite make sense to have the user code filter at runtime, so I would personally use a different directory, for example `src/test/native-image/...`

            userCodeFilterPath.set("src/test/resources/metadata.with.starting.user-code-filter/user-code-filter.json")
        }
        direct {
            options.add("config-output-dir=src/test/resources/direct-mode-metadata")
            options.add("experimental-configuration-with-origins")
        }
    }
}
----

If you want to **enable the agent through the command line**, you can specify in which mode you want to run it.
For example
[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----
./gradlew -Pagent=standard nativeTest
./gradlew -Pagent=conditional nativeTest
./gradlew -Pagent=direct nativeTest
----

[[common-agent-options]]
==== Common agent options

All the mentioned modes shares certain common configuration options like:

- callerFilterFiles
- accessFilterFiles
- builtinCallerFilter
- builtinHeuristicFilter
- enableExperimentalPredefinedClasses
- enableExperimentalUnsafeAllocationTracing
- trackReflectionMetadata

[WARNING]
**These options are for advanced usages, and you can read more about them https://www.graalvm.org/latest/reference-manual/native-image/metadata/AutomaticMetadataCollection/#agent-advanced-usage[here]**.

> Complete example of the agent block should look like this:

Reviewer: ```suggestion
A complete example of the agent block should look like this:
```

    builtinHeuristicFilter.set(true)
    enableExperimentalPredefinedClasses.set(false)
    enableExperimentalUnsafeAllocationTracing.set(false)
    trackReflectionMetadata.set(true)

    metadataCopy {
        inputTaskNames.add("test")
        outputDirectories.add("src/main/resources/META-INF/native-image/<groupId>/<artifactId>/")
        mergeWithExisting.set(true)
    }
}
----

[[agent-filter-file]]
=== Reduce the amount of generated metadata

In some cases agent may include more metadata than it is actually needed. You can filter metadata using the agent filter files.
These filter files that agent consumes have the following structure:

[source,json,subs="verbatim,attributes", role="multi-language-sample"]
----
{
 "rules": [
    {"includeClasses": "some.class.to.include.**"},
    {"excludeClasses": "some.class.to.exclude.**"},
  ],
  "regexRules": [
    {"includeClasses": "regex\.example\.class.*"},
    {"excludeClasses": "regex\.example\.exclude[0-9]+"},
  ]
}
----

The process how you can pass the config files to the agent is described in the <<additional-agent-options,previous section>>.

We can see on the example how different filter files affect generated metadata:

> We are starting with the simple filter file:

Reviewer: ```suggestion
Let's start with the simple filter file:
```

[source,json,subs="verbatim,attributes", role="multi-language-sample"]
----
{
  "rules": [
    {"includeClasses": "**"}
  ]
}
----

> This filter file will instruct the agent to include everything and therefore, you will get a massive config file.

Reviewer: ```suggestion
This filter file will instruct the agent to include everything, which will result in a massive configuration file.
```

        },
        "type": "org.example.NativeTests"
      },
      {
        "condition": {
          "typeReached": "org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor"
        },
        "type": "org.example.NativeTests",
        "allDeclaredFields": true
      },
      {
        "condition": {
          "typeReached": "org.junit.jupiter.engine.descriptor.ExtensionUtils"
        },
        "type": "org.example.NativeTests"
      },
      ...
    ],
    "resources": [
      {
        "condition": {
          "typeReached": "org.junit.platform.launcher.core.LauncherFactory"
        },
        "glob": "META-INF/services/org.junit.platform.engine.TestEngine"
      },
      {
        "condition": {
          "typeReached": "java.lang.ClassLoader"
        },
        "glob": "TestResource.txt"
      },
      ...
    ],
    "bundles": [],
    "jni": [
      {
        "condition": {
          "typeReached": "java.net.InetAddress"
        },
        "type": "java.lang.Boolean",
        "methods": [
          {
            "name": "getBoolean",
            "parameterTypes": [
              "java.lang.String"
            ]
          }
        ]
      }
    ]
  }
----

> As you can see, there are lots of entries that you may don't want.

Reviewer: ```suggestion
As you can see, there are lots of entries that you likely don't want.
They are present because the metadata was generated using tests, for example, and that the testing library was present at runtime during the agent execution pass, but in practice, your application will not use the testing library in production.
```

To reduce the amount of generated metadata, we will use the following `user-code-filter.json`:

[source,json,subs="verbatim,attributes", role="multi-language-sample"]
----
{
  "rules": [
    {"includeClasses": "**"},
    {"excludeClasses": "org.junit.**"},
    {"excludeClasses": "org.gradle.**"},
    {"excludeClasses": "worker.org.gradle.**"},
    {"excludeClasses": "org.slf4j.**"},
    {"excludeClasses": "java.**"}
  ]
}
----

[WARNING]
====

> Be careful to remove only redundant metadata and keep metadata necessary for your project.

Reviewer: ```suggestion
Always be careful when removing entries in metadata, as this may result in a broken binary.
```

====

> After we regenerate the metadata with the new filter, `reachability-metadata.json` generated on the same example as above will look like this:

Reviewer: ```suggestion
After updating our filters, we can regenerate the metadata, which will result in the following `reachability-metadata.json` file:
```

[source,json,subs="verbatim,attributes", role="multi-language-sample"]
----
{
  "reflection": [
    {
      "condition": {
        "typeReached": "org.example.NativeTests"
      },
      "type": "org.example.NativeTests$Person",
      "allDeclaredFields": true
    },
    {
      "condition": {
        "typeReached": "sun.security.jca.GetInstance"
      },
      "type": "sun.security.provider.SHA",
      "methods": [
        {
          "name": "<init>",
          "parameterTypes": []
        }
      ]
    }
  ],
  "resources": [
    {
      "condition": {
        "typeReached": "org.example.NativeTests"
      },
      "glob": "TestResource.txt"
    }
  ],
  "bundles": []
}
----

As you can see there are no more entries that contain classes from `org.junit` (as their condition) for example.

[[maintain-generated-metadata]]
== Maintain generated metadata

> If you are a library maintainer, or your application became huge, you may consider covering most of your code with tests.

Reviewer: ```suggestion
If you are a library maintainer, or if your application grows large, you may consider covering most of your code with tests.
```

This way you can also track if your project requires updates of the existing metadata.

> Considering that you run your tests in certain CI, at some point (after updating some dependency or adding new feature/test) you may notice some test failures with missing resources errors or that something is reflectively missing...

Reviewer: ```suggestion
Considering that you run your tests on CI, at some point (after updating some dependency or adding new feature/test) you may notice some test failures with missing resources errors or that something is reflectively missing...
```

In that case, your metadata requires update.

[WARNING]
Be aware that if you modified existing metadata file(s) on the default location, generating a new metadata will overwrite the existing one and your manual changes will be lost.

So if you modified existing metadata file(s) on the default location, please do the following:

> 1. Move your metadata to some other, non-default location (with the `metadataCopy` task for example)

Reviewer: :question: why is this step required if `mergeWithExisting` is true?

Author: Because we are executing `metadataCopy` as a separate task. When you run the agent, it will always overwrite existing metadata on the default location. By the time you want to execute `metadataCopy` (even with `mergeWithExisting` set to true) your manual changes are lost.

To avoid this:

-  you should copy metadata onto the "safe" location
- generate new metadata that will overwrite the original metadata
- then, when you execute `metadataCopy` with `mergeWithExisting` set to true, your new metadata will be merged with the old metadata placed on the "safe" location

Reviewer: > When you run the agent, it will always overwrite existing metadata on the default location

It shouldn't. The agent will output the metadata in a temporary directory always. It shouldn't matter how `metadataCopy` is configured. If it's the case, then it's a bug.

Author: There is a misunderstanding then... When I say **default location**, in this case, I mean `temporary directory` which is the default location where agent prints metadata if you don't do anything additionally. Maybe I should assume that at this point users already moved metadata to non-temporary (or _"non-default"_) location?

Author: Maybe I should also add some kind of tip in the `metadataCopy` section that it is strongly recommended to move metadata from the temporary directory to standard location

Author: I guess it makes more sense now?

Reviewer: Right, the confusion came from the fact I assumed that the metadata was already copied. I think we should recommend this indeed.

2. Set `mergeWithExisting` option to true in the `metadataCopy` block
3. Run your tests again to generate new metadata (as we already described in the <<collect-metadata, collect metadata section>>)
4. Run `metadataCopy` task again (with `mergeWithExisting` set to true in step 2)

This way you will keep your original metadata, and add a new one.

[[reachability-metadata-repository]]
== Reachability metadata repository

> Native Build Tools (both Gradle and Maven plugins) picks metadata from Reachability metadata repository to ensure your application works out-of-box (if all metadata required by your app is already contributed to the metadata repository).

Reviewer: ```suggestion
While we have described how you can add metadata for your own code, Native Build Tools (both Gradle and Maven plugins) makes use of metadata from Reachability Metadata Repository to ensure your application works out-of-box with 3rd party libraries that you depend on.
```

> Furthermore, you can configure Reachability metadata support through `metadataRepository` block added to our main `graalvmNative` block inside `build.gradle`.

Reviewer: ```suggestion
You can configure Reachability metadata support through `metadataRepository` block added to our main `graalvmNative` block inside `build.gradle`.
```

Most common options you may want to configure in this block are:

* `enabled` - determines if you want to use Reachability metadata support or not (`true` by default)
* `version` - specifies exact Reachability metadata version you want to use

> * `uri` - specifies the url where the metadata is stored. This can be used to point to the local repository

Reviewer: ```suggestion
* `uri` - specifies the url where the metadata is stored. This can be used to point to the local repository. If not set, it defaults to metadata which is shipped inside the Native Build Tools.
```

> YYou can read more about __Reachability metadata support__ and other (advanced) configuring options, https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html#_configuring_the_metadata_repository[here].

Reviewer: ```suggestion
You can read more about __Reachability metadata support__ and other (advanced) configuring options, https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html#_configuring_the_metadata_repository[here].
```

In some cases, when you want to maintain multiple projects that share common metadata from various libraries, you should consider contributing metadata to https://github.com/oracle/graalvm-reachability-metadata[Reachability metadata project].
https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md[Contributing to the repository] should be simple:

- Clone repository locally:

[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----
git clone git@github.com:oracle/graalvm-reachability-metadata.git
----

- generate metadata and test stubs (replace with the GAV coordinates of library you are providing metadata for):

[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----
./gradlew scaffold --coordinates com.example:my-library:1.0.0
----

- implement tests in test stubs that will show how you have generated metadata
- collect metadata as described https://github.com/oracle/graalvm-reachability-metadata/blob/master/docs/CollectingMetadata.md#collecting-metadata-for-a-library[here]
- create a pull request and fill the checklist

[[track-diagnostics]]

> == Track diagnostics

Reviewer: ```suggestion
== Using diagnostics
```

If you want to explore details about native images you are generating, you can add:

[source,bash,subs="verbatim,attributes", role="multi-language-sample"]
----

> buildArgs.add("--emit build-report")

Reviewer: :?: if it is really `--emit build-report`, then it should be 2 build args: `--emit`, `build-report`.

Author: Of course we can also separate this and it will work, but in my opinion it is more understandable if we group argument with its value. It will work both ways.

----

> For GraalVM versions starting from the GraalVM for JDK23

Reviewer: ```suggestion
For GraalVM versions starting from GraalVM for JDK 23
```

  id("org.graalvm.buildtools.native") version "{gradle-plugin-version}"
----

Replace `{maven-plugin-version}` with the latest released version.
All plugin versions are listed https://github.com/graalvm/native-build-tools/releases[here].

> - Provide the application main class in _build.gradle_ / _build.gradle.kts_ as shown below.

Reviewer 2: @dnestoro, is it a must-have step to enable the Gradle plugin? Also, is it still required to  declare an additional plugin repository in settings.gradle? 

```
 pluginManagement {
     repositories {
         mavenCentral()
         gradlePluginPortal()
     }
 }
```

Author: No, I was able to run few reproducers without that.
